File: truss-compiler/src/main/java/truss/compiler/Message.java
package truss.compiler;

import org.antlr.runtime.IntStream;
import org.antlr.runtime.RecognitionException;
import org.apache.commons.lang.Validate;
import truss.compiler.parser.TrussLexer;
import truss.compiler.parser.TrussParser;

public class Message {
    public static final String ERROR_CODE_APPLICATION = "DTL";

    private final MessageType type;
    private final Span span;
    private final String message;

    public Message(MessageType type, Object... args) {
        this(type, Span.EMPTY, args);
    }

    public Message(MessageType type, String fileName, Object... args) {
        this(type, new Span(fileName), args);
    }

    public Message(MessageType type, Span span, Object... args) {
        Validate.notNull(type, "type");
        Validate.notNull(span, "span");

        this.type = type;
        this.span = span;

        StringBuilder sb = new StringBuilder();

        sb.append(span.toString());
        sb.append(": ");
        sb.append(type.getSeverity().getCode());
        sb.append(' ');
        sb.append(ERROR_CODE_APPLICATION);
        sb.append(String.format("%04d", type.getNumber()));
        sb.append(": ");

        if (args != null && args.length > 0) {
            sb.append(String.format(type.getMessage(), args));
        } else {
            sb.append(type.getMessage());
        }

        message = sb.toString();
    }

    public MessageType getType() {
        return type;
    }

    public Span getSpan() {
        return span;
    }

    public String getMessage() {
        return message;
    }

    @Override
    public String toString() {
        return message;
    }

    public static Message fromRecognitionException(String fileName, RecognitionException e) {
        Span span = new Span(fileName, e.line, e.charPositionInLine);

        if (e.token != null && e.token.getType() == TrussLexer.EOF) {
            return new Message(MessageType.UNEXPECTED_EOF, span);
        }

        return new Message(
            MessageType.UNEXPECTED_SYNTAX,
            span,
            e.token == null ? "" : TrussParser.tokenNames[e.token.getType()]
        );
    }

    public static Message fromMismatchedToken(Span span, IntStream input, int ttype) {
        if (ttype == TrussLexer.EOF) {
            return new Message(MessageType.UNEXPECTED_EOF, span);
        }

        return new Message(MessageType.EXPECTED_SYNTAX, span, TrussParser.tokenNames[ttype]);
    }

    public static Message fromException(Exception exception) {
        if (exception instanceof RecognitionException) {
            return fromRecognitionException(null, (RecognitionException)exception);
        }
        return new Message(MessageType.UNKNOWN, exception.getMessage());
    }
}

File: Truss.Compiler/Message.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Antlr.Runtime;
using Truss.Compiler.Parser;

namespace Truss.Compiler {
    public class Message {
        public const string ErrorCodeApplication = "TRS";

        public MessageType Type { get; private set; }

        public Span Span { get; private set; }

        public string Text { get; private set; }

        public Message(MessageType type, params object[] args)
            : this(type, Span.Empty, args) {
        }

        public Message(MessageType type, string fileName, params object[] args)
            : this(type, new Span(fileName), args) {
        }

        public Message(MessageType type, Span span, params object[] args) {
            if (type == null) {
                throw new ArgumentNullException("type");
            }
            if (span == null) {
                throw new ArgumentNullException("span");
            }

            Type = type;
            Span = span;

            var sb = new StringBuilder();

            sb.Append(span);
            sb.Append(": ");
            sb.Append(type.Severity.ToString().ToLower());
            sb.Append(' ');
            sb.Append(ErrorCodeApplication);
            sb.Append(string.Format("{0:0000}", type.Number));
            sb.Append(": ");

            if (args != null && args.Length > 0) {
                sb.Append(string.Format(type.Message, args));
            } else {
                sb.Append(type.Message);
            }

            Text = sb.ToString();
        }

        public override string ToString() {
            return Text;
        }

        public static Message FromRecognitionException(string fileName, RecognitionException e) {
            var span = new Span(fileName, e.Line, e.CharPositionInLine);

            if (e.Token != null && e.Token.Type == TrussLexer.EOF) {
                return new Message(MessageType.UnexpectedEof, span);
            }

            return new Message(
                MessageType.UnexpectedSyntax,
                span,
                e.Token == null ? "" : TrussParser.tokenNames[e.Token.Type]
            );
        }

        public static Message FromMismatchedToken(Span span, IIntStream input, int ttype) {
            if (ttype == TrussLexer.EOF) {
                return new Message(MessageType.UnexpectedEof, span);
            }

            return new Message(MessageType.ExpectedSyntax, span, TrussParser.tokenNames[ttype]);
        }

        public static Message FromException(Exception exception) {
            if (exception is RecognitionException) {
                return FromRecognitionException(null, (RecognitionException)exception);
            }

            return new Message(MessageType.Unknown, exception.Message);
        }
    }
}
